src/get-or-create-render-context.js

```diff
@@ -55,62 +52,43 @@ const createRenderContext = function(jsPackages, pathToReactComponent) {
     // This makes sure that qTip2 doesn't try to use the canvas.
     sandbox.HTMLCanvasElement.prototype.getContext = undefined;
 
-    let cumulativePackageSize = 0;
-
     const context = vm.createContext(sandbox);
 
-    jsPackages.forEach((pkg) => {
-        vm.runInContext(pkg[1], context);   // pkg is [url path, contents]
-        cumulativePackageSize += pkg[1].length;
-    });
-
-    context.pathToReactComponent = pathToReactComponent;
+    // Setup callback.
+    // This indicates to the rendering code that it is involved in SSR
+    // and provides it the means to register for SSR to occur.
+    // This has to happen before we import the entry point, so it can
+    // invoke this method.
+    //
+    // The callback provided by the rendering code must return a promise that
+    // will perform the render, including data fetch calls using
+    // getDataFromTree.
+    //    getRenderPromiseCallback:
+    //          (props, maybeApolloClient) => Promise.resolve({html, css})
     runInContext(context, () => {
-        // Get stuff out of the context and into local vars.
-        const ReactDOMServer = global.ReactDOMServer;
-
-        const React = KAdefine.require("react");
-
-        // Verify we have the right version of react.  ReactDOMServer
-        // holds the react version that we've installed here, for this
-        // server, while React holds the react version that is
-        // provided via the input js packages -- that is, the version
-        // on webapp.  If they don't match, throw an exception.
-        if (React.version !== ReactDOMServer.version) {
-            throw new Error(`Server should be using React version ` +
-                            `${React.version}, but is using React ` +
-                            `version ${ReactDOMServer.version}`);
-        }
-
-        try {
-            global.StyleSheetServer = KAdefine.require("aphrodite").StyleSheetServer;
-
-            // Make sure we're using a new enough version of Aphrodite
-            global.StyleSheetServer.renderStatic; // eslint-disable-line no-unused-expressions
-        } catch (e) {
-            // If we're here, it should mean that the component being rendered
-            // does not depend on Aphrodite. We'll make a stub instead to make
-            // the code below simpler.
-            global.StyleSheetServer = {
-                renderStatic: (cb) => {
-                    return {
-                        html: cb(),
-                        css: {
-                            content: "",
-                            renderedClassNames: [],
-                        },
-                    };
-                },
+        window.__registerForSSR__ = getRenderPromiseCallback => {
```

Comment:
This is the only bit of server code we run in the client context when setting up the context now. All it does is capture the callback that the client-side entrypoint has registered.